app/src/main/java/fr/free/nrw/commons/contributions/ContributionController.java

```diff
@@ -46,11 +55,112 @@ public void initiateCameraPick(Activity activity) {
+     * If the user selects "Yes", then:
+     *
+     * Location is taken from the EXIF if the default camera application
+     * does not redact location tags.
+     *
+     * Otherwise, if the EXIF metadata does not have location information,
+     * then location captured by the app is used
+     *
+     * @param activity
+     */
+    private void askUserToAllowLocationAccess(Activity activity) {
+        DialogUtil.showAlertDialog(activity,
+            activity.getString(R.string.location_permission_title),
+            activity.getString(R.string.in_app_camera_location_access_explanation),
+            activity.getString(R.string.option_allow),
+            activity.getString(R.string.option_dismiss),
+            ()-> requestForLocationAccess(activity),
+            () -> initiateCameraUpload(activity),
+            null,
+            true);
+    }
+
+    /**
+     * Ask for location permission if the user agrees on attaching location with pictures
+     * and the app does not have the access to location
+     *
+     * @param activity
+     */
+
+    private void requestForLocationAccess(Activity activity) {
+        PermissionUtils.checkPermissionsAndPerformAction(activity,
+            permission.ACCESS_FINE_LOCATION,
+            () -> onLocationPermissionGranted(activity),
+            () -> {},
+            R.string.ask_to_turn_location_on,
+            R.string.in_app_camera_location_permission_rationale);
+    }
+
+    /**
+     * Check if apps have access to location even after having individual access
+     *
+     * @return
+     */
+    private boolean isLocationAccessToAppsTurnedOn() {
+        return (locationManager.isNetworkProviderEnabled() || locationManager.isGPSProviderEnabled());
+    }
+
+    /**
+     * Initiate in-app camera if apps have access to location.
+     * Otherwise, show location-off dialog.
+     *
+     * @param activity
+     */
+    private void onLocationPermissionGranted(Activity activity) {
+        if (!isLocationAccessToAppsTurnedOn()) {
+            showLocationOffDialog(activity);
+        } else {
+            initiateCameraUpload(activity);
+        }
+    }
+
+    /**
+     * Ask user to grant location access to apps
+     *
+     * @param activity
+     */
+
+    private void showLocationOffDialog(Activity activity) {
+        DialogUtil
+            .showAlertDialog(activity,
+                activity.getString(R.string.location_permission_title),
+                activity.getString(R.string.in_app_camera_needs_location),
+                activity.getString(R.string.title_app_shortcut_setting),
+                () -> openLocationSettings(activity),
+                true);
+    }
+
+    /**
+     * Open location source settings so that apps with location access can access it
+     *
+     * @param activity
+     */
+
+    private void openLocationSettings(Activity activity) {
+        final Intent intent = new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS);
```

Comment:
Strangely, this is opening the app's setting screen instead of the location settings screen on my device (Nord running Android 12). Any thoughts on possible reasons for this ? 🤔

Reply:
I observed a similar behavior.

Reply:
This is used at many places in the app already and works pretty well on my device, so I assumed it to be correct. Will check this out. Thank you for mentioning this!